cluster-autoscaler/core/scaledown/actuation/drain.go

```diff
@@ -174,54 +131,90 @@ func (e Evictor) DrainNodeWithPods(ctx *acontext.AutoscalingContext, node *apiv1
 	return evictionResults, errors.NewAutoscalerError(errors.TransientError, "Failed to drain node %s/%s: pods remaining after timeout", node.Namespace, node.Name)
 }
 
-// EvictDaemonSetPods creates eviction objects for all DaemonSet pods on the node.
-func (e Evictor) EvictDaemonSetPods(ctx *acontext.AutoscalingContext, nodeInfo *framework.NodeInfo, timeNow time.Time) error {
+// evictDaemonSetPods creates eviction objects for all DaemonSet pods on the node.
+func evictDaemonSetPods(ctx *acontext.AutoscalingContext, nodeInfo *framework.NodeInfo, deleteOptions simulator.NodeDeleteOptions,
+	dsEvictionEmptyNodeTimeout, dsEvictionRetryTime time.Duration, register evictionRegister) error {
 	nodeToDelete := nodeInfo.Node()
-	_, daemonSetPods, _, err := simulator.GetPodsToMove(nodeInfo, e.deleteOptions, nil, []*policyv1.PodDisruptionBudget{}, timeNow)
+	timeNow := time.Now()
+	_, daemonSetPods, _, err := simulator.GetPodsToMove(nodeInfo, deleteOptions, nil, []*policyv1.PodDisruptionBudget{}, timeNow)
 	if err != nil {
 		return fmt.Errorf("failed to get DaemonSet pods for %s (error: %v)", nodeToDelete.Name, err)
 	}
 
 	daemonSetPods = daemonset.PodsToEvict(daemonSetPods, ctx.DaemonSetEvictionForEmptyNodes)
+	evictionResults := make(map[string]status.PodEvictionResult)
+	_, err = initiateEviction(ctx, nodeToDelete, daemonSetPods, evictionResults, timeNow.Add(dsEvictionEmptyNodeTimeout),
+		dsEvictionEmptyNodeTimeout+100*time.Millisecond, dsEvictionRetryTime, register, int64(ctx.MaxGracefulTerminationSec), false)
+	return err
+}
 
-	dsEviction := make(chan status.PodEvictionResult, len(daemonSetPods))
+func initiateEviction(ctx *acontext.AutoscalingContext, node *apiv1.Node, pods []*apiv1.Pod, evictionResults map[string]status.PodEvictionResult,
```

Comment:
This one is especially hard to read because of the DS/non-DS differences, and looks like it's mostly because we kept the flow of the previous logic. I'd try rewriting it without paying attention to the previous flow, just keeping the semantics and trying to find a more readable solution.

Reply:
Maybe try to extract logic that is only applicable to fullEviction/bestEffort pods to dedicated functions that would only be called for one of the lists?

Reply:
Done